modules/photo/src/contrast_preserve.hpp

```diff
@@ -81,350 +83,331 @@ int round_num(double a)
 {
-    vector <double> initRGB;
-
-    initRGB.push_back( .33 );
-    initRGB.push_back( .33 );
-    initRGB.push_back( .33 );
-    wei = product(comb,initRGB);
+    double initRGB[3] = { .33, .33, .33 };
+    wei = product(comb, initRGB);
 
-    vector <int> sum;
+    vector<int> sum(comb.size());
 
-    for(unsigned int i=0;i<comb.size();i++)
-        sum.push_back(comb[i][0] + comb[i][1] + comb[i][2]);
+    for (unsigned int i = 0; i<comb.size(); i++)
+        sum[i] = (comb[i][0] + comb[i][1] + comb[i][2]);
 
-    for(unsigned int i=0;i<sum.size();i++)
+    for (unsigned int i = 0; i<sum.size(); i++)
     {
-        if(sum[i] == 1)
+        if (sum[i] == 1)
             wei[i] = wei[i] * double(1);
         else
             wei[i] = wei[i] * double(0);
     }
-
-    initRGB.clear();
-    sum.clear();
-
 }
 
-void Decolor::grayImContruct(vector <double> &wei, Mat img, Mat &Gray)
+void Decolor::grayImContruct(const vector<double> &wei, const Mat &img, Mat &Gray) const
 {
-    int h=img.size().height;
-    int w=img.size().width;
+    const int h = img.rows;
+    const int w = img.cols;
 
-    vector <Mat> rgb_channel;
-    split(img,rgb_channel);
+    Mat rgb_channel[3];
+    split(img, rgb_channel);
 
-    int kk =0;
+    int kk = 0;
 
-    for(int r =0;r<=order;r++)
-        for(int g=0;g<=order;g++)
-            for(int b=0;b<=order;b++)
-                if((r + g + b) <=order && (r+g+b) > 0)
+    for (int r = 0; r <= order; r++)
+        for (int g = 0; g <= order; g++)
+            for (int b = 0; b <= order; b++)
+                if ((r + g + b) <= order && (r + g + b) > 0)
                 {
-                    for(int i = 0;i<h;i++)
-                        for(int j=0;j<w;j++)
-                            Gray.at<float>(i,j)=Gray.at<float>(i,j) +
-                                (float) wei[kk]*pow(rgb_channel[2].at<float>(i,j),r)*pow(rgb_channel[1].at<float>(i,j),g)*
-                                pow(rgb_channel[0].at<float>(i,j),b);
-
-                    kk=kk+1;
+                    for (int i = 0; i < h; i++)
+                        for (int j = 0; j < w; j++)
+                            Gray.at<float>(i, j) =
+                            Gray.at<float>(i, j) +
+                            static_cast<float>(wei[kk]) *
+                            pow(rgb_channel[2].at<float>(i, j), r)*
+                            pow(rgb_channel[1].at<float>(i, j), g)*
+                            pow(rgb_channel[0].at<float>(i, j), b);
+
+                    kk = kk + 1;
                 }
 
     float minval = FLT_MAX;
     float maxval = -FLT_MAX;
 
-    for(int i=0;i<h;i++)
-        for(int j =0;j<w;j++)
-       {
-            if(Gray.at<float>(i,j) < minval)
-                minval = Gray.at<float>(i,j);
+    for (int i = 0; i<h; i++)
+        for (int j = 0; j<w; j++)
+        {
+            if (Gray.at<float>(i, j) < minval)
+                minval = Gray.at<float>(i, j);
```

Comment:
It is shorter:
```
float v = Gray.at<float>(i, j);
minval = std::min(minval, v);
maxval = std::max(maxval, v);
```

Or better to use well optimized [`cv::minmaxLoc()`](https://docs.opencv.org/master/d2/de8/group__core__array.html#gab473bf2eb6d14ff97e89b355dac20707) instead of these loops.

Reply:
👍   I will try this change in the next refactoring patch.